http-server-netty/src/main/java/io/micronaut/http/server/netty/NativeTransportAvailabilityCondition.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2017-2019 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.server.netty;
+
+import java.util.Optional;
+
+import io.micronaut.context.BeanContext;
+import io.micronaut.context.condition.Condition;
+import io.micronaut.context.condition.ConditionContext;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.reflect.ClassUtils;
+import io.micronaut.core.reflect.ReflectionUtils;
+
+/**
+ * Checks if netty's native transport is available.
+ * @author croudet
+ */
+@Internal
+class NativeTransportAvailabilityCondition implements Condition {
+    private final String className;
+
+    /**
+     * Checks if netty's native transport is available.
+     * @param className The class name to check.
+     */
+    NativeTransportAvailabilityCondition(String className) {
+        this.className = className;
+    }
+
+    /**
+     * Checks if netty's native transport is available.
+     * @return true if the specified native transport class is on the classpath and is 'isAvailable()' is true.
+     */
+    @Override
+    public boolean matches(ConditionContext context) {
+        BeanContext beanContext = context.getBeanContext();
+        Optional<Class> nativeTransport = ClassUtils.forName(className, beanContext.getClassLoader());
+        Optional<Boolean> nativeTransportAvailable = nativeTransport.flatMap(cl -> ReflectionUtils.findMethod(cl, "isAvailable")).flatMap(method -> Optional.of(ReflectionUtils.invokeMethod(null, method)));
```

Comment:
This should be done without reflection

Reply:
Ok, if I use
```java
/**
 * Factory for KQueueEventLoopGroup.
 * @author croudet
 */
@Factory
@Requires(classes= KQueue.class)
@Requires(condition = KQueueAvailabilityCondition.class)
@Requires(property = "micronaut.server.netty.use.native.transport", value = StringUtils.TRUE, defaultValue = StringUtils.FALSE)
@Internal
class KQueueEventLoopGroupFactory implements EventLoopGroupFactory {
```
with `KQueueAvailabilityCondition`
```java
...
    /**
     * Checks if netty's native transport is available.
     * @return true 'KQueue.isAvailable()' is true.
     */
    @Override
    public boolean matches(ConditionContext context) {
      return KQueue.isAvailable();
    }
...
```
is `@Requires(classes= KQueue.class)` evaluated before '@Requires(condition = KQueueAvailabilityCondition.class)`?

Reply:
If you use a single annotation, you can be sure the class is checked first. `@Requires(classes= KQueue.class, condition = KQueueAvailabilityCondition.class)`

Reply:
No longer using reflection.